lib/cache.ts

```diff
@@ -0,0 +1,44 @@
+/**
+ * Basic implementation of a cache, mostly for use in caching database results.
+ * This is explicitly made to be synchronously accessed (to pair with results from asynchronous databases.)
+ * @author mia-pi-git
+ */
+
+const CACHE_EXPIRY_TIME = 5 * 60 * 1000;
+
+export class Cache<T> {
+	readonly cache: {[k: string]: {data: T, lastCache: number}};
+	expiryTime: number;
+	dataFetcher: (key: string) => T | Promise<T>;
+	constructor(fetcher: (key: string) => T | Promise<T>, invalidateTime = CACHE_EXPIRY_TIME) {
+		this.cache = {};
+		this.expiryTime = invalidateTime;
+		this.dataFetcher = fetcher;
+	}
+	// todo make this only return T
+	// <T>roubling it doesn't, since that was the original intent, but for now this will do
```

Comment:
Shouldn't this return `Promise<T>` (but the promise will resolve immediately if it's pulled from cache)?

Reply:
No, it needs to be explicitly synchronous. The whole point of this is to not have to propagate awaits/Promises.

Reply:
Okay...
```ts
while (!variable) {
    variable = cache.get(key);
}
```
seems like a lot of jank for retrieving from the cache compared to just propagating a promise (which will resolve immediately some of the time anyway).

Reply:
I explained in dms, but I'll post for posterity. This is explicitly made so that no async/await is used, so we don't hold up on PMs.